org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/util/CharDeduplication.java

```diff
@@ -118,167 +92,60 @@ public void reset() {
+				// Successfully deduplicated:
 				return charArray;
+			}
 		}
-		// ---------other side---------
-		i = -1;
-		int max = this.newEntry4;
-		while (++i <= max) {
-			char[] charArray = table[i];
-			if ((c0 == charArray[0]) && (c1 == charArray[1]) && (c2 == charArray[2]) && (c3 == charArray[3]))
-				return charArray;
+		char[] r = Arrays.copyOfRange(source, from, to);
+		// not found -> overwrite existing entries in a circular buffer:
+		if (positionToReplace == -1) {
+			// no empty entry found - normal case:
+			int j = this.circularBufferPointer[circularBufferStart]++;
+			positionToReplace = (circularBufferStart + (j & (SEARCH_SIZE-1))) & (TABLE_SIZE - 1);
 		}
-		// --------add the entry-------
-		if (++max >= INTERNAL_TABLE_SIZE)
-			max = 0;
-		char[] r;
-		System.arraycopy(src, start, r = new char[4], 0, 4);
-		return table[this.newEntry4 = max] = r;
+		this.hashTable[positionToReplace] = r;
+		return r;
 	}
 
-	private final char[] optimizedCurrentTokenSource5(char[] source, int startPosition) {
-		char[] src = source;
-		int start = startPosition;
-		char c0, c1 = src[start + 1], c2, c3 = src[start + 3], c4;
-		int hash = (((c0 = src[start]) << 12) + ((c2 = src[start + 2]) << 6) + (c4 = src[start + 4])) % TABLE_SIZE;
-		char[][] table = this.charArray_length[3][hash];
-		int i = this.newEntry5;
-		while (++i < INTERNAL_TABLE_SIZE) {
-			char[] charArray = table[i];
-			if ((c0 == charArray[0]) && (c1 == charArray[1]) && (c2 == charArray[2]) && (c3 == charArray[3])
-					&& (c4 == charArray[4]))
-				return charArray;
+	private int hashCode(char[] source, int from, int to) {
+		int result = source[from];
+		for (int i = from + 1; i < to; i++) {
+			result = 31 * result + source[i];
 		}
-		// ---------other side---------
-		i = -1;
-		int max = this.newEntry5;
-		while (++i <= max) {
-			char[] charArray = table[i];
-			if ((c0 == charArray[0]) && (c1 == charArray[1]) && (c2 == charArray[2]) && (c3 == charArray[3])
-					&& (c4 == charArray[4]))
-				return charArray;
-		}
-		// --------add the entry-------
-		if (++max >= INTERNAL_TABLE_SIZE)
-			max = 0;
-		char[] r;
-		System.arraycopy(src, start, r = new char[5], 0, 5);
-		return table[this.newEntry5 = max] = r;
+		return result;
 	}
 
-	private final char[] optimizedCurrentTokenSource6(char[] source, int startPosition) {
-		char[] src = source;
-		int start = startPosition;
-		char c0, c1 = src[start + 1], c2, c3 = src[start + 3], c4, c5 = src[start + 5];
-		int hash = (((c0 = src[start]) << 12) + ((c2 = src[start + 2]) << 6) + (c4 = src[start + 4])) % TABLE_SIZE;
-		char[][] table = this.charArray_length[4][hash];
-		int i = this.newEntry6;
-		while (++i < INTERNAL_TABLE_SIZE) {
-			char[] charArray = table[i];
-			if ((c0 == charArray[0]) && (c1 == charArray[1]) && (c2 == charArray[2]) && (c3 == charArray[3])
-					&& (c4 == charArray[4]) && (c5 == charArray[5]))
-				return charArray;
+	private boolean equals(char[] source, int from, int to, char[] charArray) {
```

Comment:
Small thing: I suggest to use `java.util.Arrays.equals(char[], int, int, char[], int, int)` instead. It delegates to the vectorized ArraySupport.mismatch.

Reply:
i did not know that function, but it's slower - probably because tokens are typically small and i use several conditions, that don't need to be checked.
try
`Arrays.equals(source, from, to, charArray, 0,  charArray.length)`
with CharDeduplicationTest